File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"
#include<vector>
#include<string>
#include<unordered_map>

namespace protein_translation {

    std::vector<std::string> proteins(std::string sequence) {
    
        // init dictionary
        std::unordered_map<std::string, std::string> proteinMap{};
        proteinMap["AUG"] = "Methionine";
        proteinMap["UUU"] = "Phenylalanine";
        proteinMap["UUC"] = "Phenylalanine";
        proteinMap["UUA"] = "Leucine";
        proteinMap["UUG"] = "Leucine";
        proteinMap["UCU"] = "Serine";
        proteinMap["UCC"] = "Serine";
        proteinMap["UCA"] = "Serine";
        proteinMap["UCG"] = "Serine";
        proteinMap["UAU"] = "Tyrosine";
        proteinMap["UAC"] = "Tyrosine";
        proteinMap["UGU"] = "Cysteine";
        proteinMap["UGC"] = "Cysteine";
        proteinMap["UGG"] = "Tryptophan";
        proteinMap["UAA"] = "STOP";
        proteinMap["UAG"] = "STOP";
        proteinMap["UGA"] = "STOP";

        // algorithm
        std::vector<std::string> result{};
        for (size_t i = 0; i < sequence.size(); i+=3) {
            std::string slice = sequence.substr(i, 3);

            if (proteinMap[slice] == "STOP") {
                break;
            } else {
                result.push_back(proteinMap[slice]);
            }
        }

        return result;
    }    

}  // namespace protein_translation

File: solutions/go/protein-translation/1/protein_translation.go
package proteintranslation

import "errors"

var ErrInvalidBase = errors.New("no valid condon found")
var ErrStop = errors.New("encountered STOP condon")

var codonMap = map[string]string{
	"AUG": "Methionine",
	"UUU": "Phenylalanine",
	"UUC": "Phenylalanine",
	"UUA": "Leucine",
	"UUG": "Leucine",
	"UCU": "Serine",
	"UCC": "Serine",
	"UCA": "Serine",
	"UCG": "Serine",
	"UAU": "Tyrosine",
	"UAC": "Tyrosine",
	"UGU": "Cysteine",
	"UGC": "Cysteine",
	"UGG": "Tryptophan",
	"UAA": "STOP",
	"UAG": "STOP",
	"UGA": "STOP",
}

func FromRNA(rna string) ([]string, error) {
	result := []string{}
	input := []rune(rna)

	for i := 0; i < len(input); i += 3 {
		end := min(i+3, len(input))
		chunk := string(input[i:end])
		codon, err := FromCodon(chunk)
		if err != nil {
			if err == ErrInvalidBase {
				return result, err
			} else if err == ErrStop {
				break
			}
		} else {
			result = append(result, codon)
		}
	}

	return result, nil
}

func FromCodon(codon string) (string, error) {
	input := []rune(codon)
	result := ""

	for i := 0; i < len(input); i += 3 {
		end := min(i+3, len(input))
		chunk := string(input[i:end])
		value, exist := codonMap[chunk]
		if exist {
			result = value
		}
	}
	switch result {
	case "":
		return "", ErrInvalidBase
	case "STOP":
		return "", ErrStop
	default:
		return result, nil
	}
}
